src/Windows/Avalonia.Win32/DragSource.cs

```diff
@@ -8,17 +8,26 @@ namespace Avalonia.Win32
 {
     class DragSource : IPlatformDragSource
     {
-        public Task<DragDropEffects> DoDragDrop(PointerEventArgs triggerEvent,
+        public unsafe Task<DragDropEffects> DoDragDrop(PointerEventArgs triggerEvent,
             IDataObject data, DragDropEffects allowedEffects)
         {
             Dispatcher.UIThread.VerifyAccess();
+
             triggerEvent.Pointer.Capture(null);
-            OleDragSource src = new OleDragSource();
-            DataObject dataObject = new DataObject(data);
-            int allowed = (int)OleDropTarget.ConvertDropEffect(allowedEffects);
+            
+            using var dataObject = new DataObject(data);
+            using var src = new OleDragSource();
+            var allowed = OleDropTarget.ConvertDropEffect(allowedEffects);
+            
+            var objPtr = MicroCom.MicroComRuntime.GetNativeIntPtr<Win32Com.IDataObject>(dataObject);
+            var srcPtr = MicroCom.MicroComRuntime.GetNativeIntPtr<Win32Com.IDropSource>(src);
+
+            UnmanagedMethods.DoDragDrop(objPtr, srcPtr, (int)allowed, out var finalEffect);
+            
+            // Force releasing of internal wrapper to avoid memory leak, if drop target keeps com reference.
+            dataObject.ReleaseWrapped();
```

Comment:
DoDragDrop might not release COM reference, if drop target will keep it.
Qt does the same with releasing internal data object - https://github.com/qt/qtbase/blob/ced47a05aa3a59710ef1abe1474b76ddb5460049/src/plugins/platforms/windows/qwindowsdrag.cpp#L718